notion-server/src/ethereum/mod.rs

```diff
@@ -12,55 +12,129 @@ pub type Result<V, E = Error> = std::result::Result<V, E>;
 /// Errors arising from the simulation.
 #[derive(Debug, Snafu)]
 pub enum Error {
+    Decode {
+        backtrace: Backtrace,
+        source: base64::DecodeError,
+    },
+    OutOfBounds {
+        message: String,
+    },
     /// Operation was cancelled because the simulation is shutting down.
     Terminated,
 }
 
 #[derive(Debug)]
 enum Operation {
-    CreateExecutionEnvironment(args::CreateExecutionEnvironment, Sender<EeIndex>),
+    CreateExecutionEnvironment(args::CreateExecutionEnvironment, Sender<Result<u32>>),
+    CreateShardBlock(args::CreateShardBlock, Sender<Result<u32>>),
     CreateShardChain(args::CreateShardChain, Sender<u32>),
-    CreateShardBlock(args::CreateShardBlock, Sender<u32>),
-    GetShardBlock(args::GetShardBlock, Sender<ShardBlock>),
+    GetExecutionEnvironment(args::GetExecutionEnvironment, Sender<Result<args::ExecutionEnvironment>>),
+    GetShardBlock(args::GetShardBlock, Sender<Result<args::ShardBlock>>),
+    GetSimulationState(args::GetSimulationState, Sender<args::SimulationState>),
 }
 
-#[derive(Debug, Clone)]
```

Comment:
Removing `Clone` is going to cause problems from the Rocket side. [`State`](https://api.rocket.rs/v0.4/rocket/struct.State.html) only implements `Deref`, and not `DerefMut`, and [`Sender::send`](https://docs.rs/tokio/0.2.1/tokio/sync/mpsc/struct.Sender.html#method.send) requires a mutable reference.

Being able to `clone` the `Handle` is how I get around that limitation.